client/gutenberg/extensions/map/component.js

```diff
@@ -0,0 +1,338 @@
+		const { api_key, children, points, map_style, map_details } = this.props;
+		const { map } = this.state;
+		if ( api_key && api_key.length > 0 && api_key !== prevProps.api_key ) {
+			window.mapboxgl = null;
+			this.loadMapLibraries();
+		}
+		// If the user has just clicked to show the Add Point component, hide info window.
+		// AddPoint is the only possible child.
+		if ( children !== prevProps.children && children !== false ) {
+			this.clearCurrentMarker();
+		}
+		if ( points !== prevProps.points ) {
+			this.setBoundsByMarkers();
+		}
+		if ( points.length !== prevProps.points.length ) {
+			this.clearCurrentMarker();
+		}
+		if ( map_style !== prevProps.map_style || map_details !== prevProps.map_details ) {
+			map.setStyle( this.getMapStyle() );
+		}
+	}
+	/* Event handling */
+	onMarkerClick = marker => {
+		const { onMarkerClick } = this.props;
+		this.setState( { activeMarker: marker } );
+		onMarkerClick();
+	};
+	onMapClick = () => {
+		this.setState( { activeMarker: null } );
+	};
+	clearCurrentMarker = () => {
+		this.setState( { activeMarker: null } );
+	};
+	updateActiveMarker = updates => {
+		const { points } = this.props;
+		const { activeMarker } = this.state;
+		const { index } = activeMarker.props;
+		const newPoints = points.slice( 0 );
+
+		assign( newPoints[ index ], updates );
+		this.props.onSetPoints( newPoints );
+	};
+	deleteActiveMarker = () => {
+		const { points } = this.props;
+		const { activeMarker } = this.state;
+		const { index } = activeMarker.props;
+		const newPoints = points.slice( 0 );
+
+		newPoints.splice( index, 1 );
+		this.props.onSetPoints( newPoints );
+		this.setState( { activeMarker: null } );
+	};
+	// Various map functions
+	sizeMap = () => {
+		const { map } = this.state;
+		const mapEl = this.mapRef.current;
+		const blockWidth = mapEl.offsetWidth;
+		const maxHeight = window.innerHeight * 0.8;
+		const blockHeight = Math.min( blockWidth * ( 3 / 4 ), maxHeight );
+		mapEl.style.height = blockHeight + 'px';
+		map.resize();
+		this.setBoundsByMarkers();
+	};
+	setBoundsByMarkers = () => {
+		const { zoom, points, onSetZoom } = this.props;
+		const { map, activeMarker, mapboxgl, zoomControl, fit_to_bounds } = this.state;
+
+		const bounds = new mapboxgl.LngLatBounds();
+		if ( ! map || ! points.length || activeMarker ) {
+			return;
+		}
+		points.forEach( point => {
+			bounds.extend( [ point.coordinates.longitude, point.coordinates.latitude ] );
+		} );
+		map.setCenter( bounds.getCenter() );
+
+		// If there are multiple points, zoom is determined by the area they cover,
+		// and zoom control is removed.
+		if ( points.length > 1 ) {
+			map.fitBounds( bounds, {
+				padding: {
+					top: 40,
+					bottom: 40,
+					left: 20,
+					right: 20,
+				},
+			} );
+			this.setState( { fit_to_bounds: true } );
+			map.removeControl( zoomControl );
+			return;
+		}
+		/* Case where points go from multiple to just one. Set zoom to an arbitrarily high level. */
+		if ( fit_to_bounds ) {
```

Comment:
I'm having a bit of trouble following when this `fit_to_bounds` state is set so I might be totally off here. :-)

Since `setBoundsByMarkers` is called from `componentDidUpdate` where we know if _"points go from multiple to just one"_, should we either have this in a separate method or skip the state for this and use something like:
```js
setBoundsByMarkers = ( { fit_to_bounds = false } ) => {
  if ( fit_to_bounds ) {
    // ...
  }
}
```

This because to me it looked like the state is used kinda as a function argument, not as a "state of the component" per-se but I might be wrong.

Also worth noting that _[State Updates May Be Asynchronous](https://reactjs.org/docs/state-and-lifecycle.html#state-updates-may-be-asynchronous): React may batch multiple `setState()` calls into a single update for performance._

Since there's a subtle `return` here in between, I don't think you hit issues with this but mentioning just in case.

Reply:
I'd like to defer a deeper dive into `setBoundsByMarkers()` until after this merges. It could definitely use some cleanup, but since it's functional I'd like to leave it for now.